java-checks/src/main/java/org/sonar/java/checks/OneExpectedCheckExceptionCheck.java

```diff
@@ -0,0 +1,180 @@
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionStatementTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TryStatementTree;
+
+@Rule(key = "S5783")
+public class OneExpectedCheckExceptionCheck extends IssuableSubscriptionVisitor {
+
+  private static final String JUNIT4_ASSERT = "org.junit.Assert";
+
+  private static final MethodMatchers JUNIT4_ASSERT_THROWS_WITH_MESSAGE = MethodMatchers.create()
+    .ofTypes(JUNIT4_ASSERT)
+    .names("assertThrows")
+    .addParametersMatcher("java.lang.String", MethodMatchers.ANY, MethodMatchers.ANY)
+    .build();
+
+  private static final MethodMatchers ALL_ASSERT_THROWS_MATCHER = MethodMatchers.create()
+    .ofTypes(JUNIT4_ASSERT, "org.junit.jupiter.api.Assertions")
+    .names("assertThrows")
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers JUNIT_FAIL_MATCHER = MethodMatchers.create()
+    .ofTypes(JUNIT4_ASSERT, "org.junit.jupiter.api.Assertions")
+    .names("fail")
+    .withAnyParameters()
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.TRY_STATEMENT, Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+
+    if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree mit = (MethodInvocationTree) tree;
+      Arguments arguments = mit.arguments();
+      if (JUNIT4_ASSERT_THROWS_WITH_MESSAGE.matches(mit)) {
+        processAssertThrowsArguments(arguments.get(1), arguments.get(2));
+      } else if (arguments.size() >= 2 && ALL_ASSERT_THROWS_MATCHER.matches(mit)) {
+        processAssertThrowsArguments(arguments.get(0), arguments.get(1));
+      }
+    } else {
+      TryStatementTree tryStatementTree = (TryStatementTree) tree;
+      if (isTryCatchIdiom(tryStatementTree)) {
+        tryStatementTree.catches().forEach(c ->
+          reportMultipleCallThrowingExceptionInTree(c.parameter().type().symbolType(), tryStatementTree.block(), c.parameter().type())
+        );
+      }
+    }
+  }
+
+  private void processAssertThrowsArguments(ExpressionTree expectedType, ExpressionTree executable) {
+    Optional<IdentifierTree> expectedException = getExpectedException(expectedType);
```

Comment:
i would rewrite this part slightly differently:
```
if (!executable.is(Tree.Kind.LAMBDA_EXPRESSION)) {
  return;
}
getExpectedException(expectedType)
  .ifPresent(exception -> reportMultipleCallThrowingExceptionInTree(exception.symbolType(), ((LambdaExpressionTree) executable).body(), exception));
```

Reply:
Definitely better.